### PR TITLE
Mock epoch index and randomness in SproofBuilder (for testing)

### PR DESCRIPTION
**Pull request by 4meta5**
Follow up for #1419 -- moonbeam integration tests require inserting a mocked epoch index and randomness into the relay chain state sproof.

This PR adds 2 fields to `MockValidationDataInherentDataProvider`:
1. `para_blocks_per_relay_epoch` to configure relay epoch index 
2. `relay_randomness_config` to configure BABE one epoch ago randomness value 

The epoch index and randomness are inserted into the relay state proof every block.

moonbeam testing PR where this code is used: https://github.com/PureStake/moonbeam/pull/1779

### CODE REVIEW

File: primitives/parachain-inherent/src/mock.rs [hunk 1]
```diff
@@ -178,6 +178,14 @@ impl InherentDataProvider for MockValidationDataInherentDataProvider {
 			sproof_builder.upsert_inbound_channel(*para_id).mqc_head = Some(channel_mqc.head());
 		}
 
+		// Epoch is set equal to current para block
+		sproof_builder.current_epoch = self.current_para_block.into();
```
Comment (bkchr) [inline primitives/parachain-inherent/src/mock.rs:182]: Should we not make this configurable to the outside? I find it weird to use the current para block for this.

Comment (4meta5) [reply primitives/parachain-inherent/src/mock.rs]: I made the epoch configurable in terms of the number of parachain blocks per relay chain epoch. Let me know if you had something else in mind.

File: primitives/parachain-inherent/src/mock.rs [hunk 2]
```diff
@@ -59,6 +61,30 @@ pub struct MockValidationDataInherentDataProvider {
...
+		mock_randomness.into()
+	}
+}
+
+/// Parameters for how the Mock inherent data provider should inject the relay
+/// epoch index and BABE one epoch ago randomness.
+pub struct MockRandomnessConfig<R> {
+	/// Number of parachain blocks per relay chain epoch
+	/// Mock epoch is computed by dividing `current_para_block` by this value.
+	pub para_blocks_per_epoch: u32,
```
Comment (bkchr) [inline primitives/parachain-inherent/src/mock.rs:83]: Why is that part of the `RandomnessConfig`?

Comment (4meta5) [reply primitives/parachain-inherent/src/mock.rs:83]: The default configuration updates the randomness whenever the epoch changes so I put them together, but now I'm rethinking it.

Comment (4meta5) [reply primitives/parachain-inherent/src/mock.rs:83]: I'll move the fields into `MockValidationDataInherentDataProvider` and get rid of this struct as that is preferred.

File: primitives/parachain-inherent/src/mock.rs [hunk 3]
```diff
@@ -178,6 +199,20 @@ impl InherentDataProvider for MockValidationDataInherentDataProvider {
...
+			0u64
+		} else if self.para_blocks_per_relay_epoch == 0u32 {
+			// do not divide by 0 => set epoch to para block number
+			self.current_para_block.into()
+		} else {
+			(self.current_para_block / self.para_blocks_per_relay_epoch).into()
+		};
+		// Randomness is set by randomness generator
+		sproof_builder.randomness =
+			self.relay_randomness_config.generate_randomness(sproof_builder.current_epoch);
```
Comment (bkchr) [inline primitives/parachain-inherent/src/mock.rs:214]:
```suggestion
			self.relay_randomness_config.generate_randomness(sproof_builder.current_para_block.into());
```

File: primitives/parachain-inherent/src/mock.rs [hunk 4]
```diff
@@ -178,6 +199,20 @@ impl InherentDataProvider for MockValidationDataInherentDataProvider {
...
+		// Epoch is set equal to current para block / blocks per epoch
+		sproof_builder.current_epoch = if self.current_para_block < self.para_blocks_per_relay_epoch
+		{
+			0u64
+		} else if self.para_blocks_per_relay_epoch == 0u32 {
+			// do not divide by 0 => set epoch to para block number
+			self.current_para_block.into()
+		} else {
+			(self.current_para_block / self.para_blocks_per_relay_epoch).into()
+		};
```
Comment (bkchr) [inline primitives/parachain-inherent/src/mock.rs:211]:
```suggestion
		sproof_builder.current_epoch = if self.para_blocks_per_relay_epoch == 0 {
			// do not divide by 0 => set epoch to para block number
			self.current_para_block.into()
		} else {
			(self.current_para_block / self.para_blocks_per_relay_epoch).into()
		};
```

**Review by bkchr — APPROVED** [review]

File: primitives/parachain-inherent/src/mock.rs [hunk 5]
```diff
@@ -51,6 +51,11 @@ pub struct MockValidationDataInherentDataProvider {
 	/// The number of relay blocks that elapses between each parablock. Probably set this to 1 or 2
 	/// to simulate optimistic or realistic relay chain behavior.
 	pub relay_blocks_per_para_block: u32,
+	/// Number of parachain blocks per relay chain epoch
+	/// Mock epoch is computed by dividing `current_para_block` by this value.
+	pub para_blocks_per_relay_epoch: u32,
+	/// Function to mock BABE one epoch ago randomness using input current epoch
```
Comment (4meta5) [inline primitives/parachain-inherent/src/mock.rs:57]:
```suggestion
	/// Function to mock BABE one epoch ago randomness
```

**Pull request merged**